hazelcast-jet-sql/src/main/java/com/hazelcast/jet/sql/impl/aggregate/AvgSqlAggregation.java

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.jet.sql.impl.aggregate;
+
+import com.hazelcast.nio.ObjectDataInput;
+import com.hazelcast.nio.ObjectDataOutput;
+import com.hazelcast.sql.impl.QueryException;
+import com.hazelcast.sql.impl.type.QueryDataType;
+import com.hazelcast.sql.impl.type.QueryDataTypeFamily;
+
+import javax.annotation.concurrent.NotThreadSafe;
+import java.io.IOException;
+import java.math.BigDecimal;
+import java.util.Objects;
+
+import static com.hazelcast.sql.impl.expression.math.ExpressionMath.DECIMAL_MATH_CONTEXT;
+
+@NotThreadSafe
+public class AvgSqlAggregation extends SqlAggregation {
+
+    private QueryDataType resultType;
+
+    private SumSqlAggregation sum;
+    private CountSqlAggregation count;
+
+    @SuppressWarnings("unused")
+    private AvgSqlAggregation() {
+    }
+
+    public AvgSqlAggregation(int index, QueryDataType operandType) {
+        this(index, operandType, false);
+    }
+
+    public AvgSqlAggregation(int index, QueryDataType operandType, boolean distinct) {
+        super(index, true, distinct);
+        this.resultType = inferResultType(operandType);
+        this.sum = new SumSqlAggregation(index, operandType);
+        this.count = new CountSqlAggregation();
+    }
+
+    private static QueryDataType inferResultType(QueryDataType operandType) {
```

Comment:
Please note that in IMDG we moved type inference to Calcite-related classes. This was a tedious task, but ultimately it gives us full control of possible conversion, type inference, error messages with the exact parser positions, etc.
For aggregations we are likely to follow a similar approach - the inference will be implemented at the Calcite level. We may have a kind of mismatch here going forward, that will require you to change the implementation if we want to have a common aggregation-related code base. This doesn't seem critical to me at the moment, since your implementation of inference is simple, so just informing you about the possible mismatch in approaches.

Reply:
We do rely on `HazelcastTypeSystem` (see `HazelcastTypeSystem#deriveSumType` & `HazelcastTypeSystem#deriveAvgAggType`), is that what you mean? The duplication here is for a reason i think - most probably I could not find a way to relay that information from Calcite - but I can't remember the details right now... I'll try to recall what was the problem.

Reply:
I tried to reuse `HazelcastTypeSystem` derive methods but because `RelDataType` is not serializable the code became quite convoluted. The duplication isn't substantial, I think we can live with that, the discrepancy should be pretty quickly apparent.